crates/conformance/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+//! Spin Conformance Tests
+//!
+//! This crate provides a suite of tests to determine which Spin functions a guest module is capable of
+//! implementing (in the case of guest-implemented functions such as triggers) or using (in the case of
+//! host-implemented functions such as outbound HTTP requests).
+//!
+//! The suite may be run via the [`test()`] function, which accepts a [`wasmtime::Module`] and a [`Config`] and
+//! returns a [`Report`] which details which features the module does or does not support.  "Support" in this
+//! context depends on whether the test is for a function implemented by the guest (e.g. triggers) or by the host
+//! (e.g. outbound requests).
+//!
+//! - For a guest-implemented function, the test will call the function and assert the result matches what is
+//! expected (see [`Report::http_trigger`] for an example).
+//!
+//! - For a host-implemented function, the test will call into the guest according to the specified
+//! [`InvocationStyle`] with a set of arguments indicating which host function to call and with what arguments.
+//! The test then asserts the host function was indeed called with the expected arguments (see [`Report::config`]
+//! for an example).
```

Comment:
Thanks, this certainly helps.  I do feel the project could do with more sense of motivation, who would use it and when, and how to get value out of it, though that could equally well go in a read-me.  The sort of thing I'm thinking of is:

* This is a way for language integrators / SDK authors to check that their integrations/SDKs work correctly with Spin ABIs.
* It is not for application authors.  The testee modules must be written specifically to handle the instructions sent by the conformance suite, and (where appropriate) to make specific outgoing calls in response.
* The way we see it being used is that a SDK author writes a module that goes through their SDK to do what the conformance test needs.  Then they can run the conformance tester with that module as they work on the SDK.  (Can we think of a term for modules written for this purpose?  "Testee" and "guest" are both very general-purpose.  "Conformance test module" maybe or does that get mixed up with the tester?)
  * I dunno if you can do this in Rust docs but a super simple boxes and arrows picture/sequence diagram (showing the interactions of the harness, the SDK and the custom guest code) might say a lot!
* The way the test harness exercises the SDK is:
  * For triggers, to call those triggers with synthetic events
  * For host functions, by sending instructions telling the testee to call a given function with given arguments
  * I know the doc comments already say this bit already but it's currently phrased very much in implementation terms and I think it would be good to soften it with a bit of the "this is the high-level shape of what it does" wording

I'd avoid phrases like "the test will call into the guest" because wait, aren't I writing the test?  (No, I'm writing the guest, but the guest module is a test module, and the real testee is the SDK.)  Hence my suggestion for specific terms for the test harness and the module that uses the SDK.

Similarly, I'd avoid phrases like "the module does or does not support" or "the guest module is capable of implementing" because it's not really about whether the _module_ supports them - it's about whether the _SDK_ supports them, and hence whether _modules built using that SDK_ are capable of implementing them.

I hope this is helpful!  I'm finding it a little bit tricky to clearly articulate where I'm sliding off the docs as currently written, partly due to the Monad Curse Effect, but I think "high level motivating overview" is really the nub of it.  Happy to chat and brainstorm further if that would help.